File: ex1/arrange/arrange.cpp
#include <vector>
#include <unordered_map>
#include <cstdio>
#include <cstdlib>
#include <cmath>

#define ll long long

using namespace std;

struct tree {
    ll val;
    tree *left, *right;
    tree (ll v) : val(v), left(nullptr), right(nullptr) {}
};

tree * mk_tree(FILE *f) {
    ll num;
    fscanf(f, "%lld", &num);
    if (num) {
        tree *t = new tree(num);
        t->left = mk_tree(f);
        t->right = mk_tree(f);
        return t;
    }
    else return nullptr;
}

ll arrange(tree *t, ll N) {
    if (t == nullptr) return N+1;
    ll l = arrange(t->left, N);
    ll r = arrange(t->right, N);
    if (l > r && l <= N && r <= N) {
            tree *tmp = t->left;
            t->left = t->right;
            t->right = tmp;
            l = r;
    }  
    else if (l == N+1 || r == N+1) {
        if ((l != N+1 && t->val < l) || (r != N+1 && t->val > r)) {
            tree *tmp = t->left;
            t->left = t->right;
            t->right = tmp;
            l = r;
        }

    }
    return l == N+1 ? t->val : l;
}

void print_tree(tree *t, bool &first) {
    if (t == nullptr) return;
    print_tree(t->left, first);
    printf(first ? "%lld" : " %lld", t->val);
    first = false;
    print_tree(t->right, first);
}

int main (int argc, char **argv) {
    if (argc != 2) {
        printf("Usage: %s <input_file>\n", argv[0]);
        return 1;
    }
    FILE *f = fopen(argv[1], "r");
    if (f == NULL) {
        printf("Error opening file\n");
        return 1;
    }
    ll N;
    fscanf(f, "%lld", &N);
    tree *t = mk_tree(f);
    fclose(f);
    arrange(t, N);
    bool first = true;
    print_tree(t, first);
    printf("\n");
}        

File: ex2/arrange/Arrange.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Stack;
import java.util.Scanner;

class Tree {
    long val;
    Tree left, right;
    long minVal = 1000001;

    Tree(long v) {
        val = v;
        left = right = null;
    }
}

public class Arrange {

    public static Tree mkTree(Scanner scanner) {
        Stack<Tree> stack = new Stack<>();
        Tree root = null, curr = null;

        while (scanner.hasNext()) {
            if (scanner.hasNextLong()) {
                long num = scanner.nextLong();
                if (num != 0) {
                    Tree newNode = new Tree(num);
                    if (curr == null) {
                        root = newNode;
                    } else if (curr.left == null) {
                        curr.left = newNode;
                    } else {
                        curr.right = newNode;
                    }
                    stack.push(newNode);
                    curr = newNode;
                } else {
                    if (!stack.isEmpty()) {
                        curr = stack.pop();
                    }
                }
            } else {
                scanner.next(); 
            }
        }

        return root;
    }

    public static void arrange(Tree root, long N) {
        if (root == null) return;
        Stack<Tree> stack1 = new Stack<>();
        Stack<Tree> stack2 = new Stack<>();
        stack1.push(root);

        while (!stack1.isEmpty()) {
            Tree node = stack1.pop();
            stack2.push(node);

            if (node.left != null) {
                stack1.push(node.left);
            }
            if (node.right != null) {
                stack1.push(node.right);
            }
        }

        while (!stack2.isEmpty()) {
            Tree node = stack2.pop();
            long l = (node.left != null) ? node.left.minVal : N + 1;
            long r = (node.right != null) ? node.right.minVal : N + 1;

            if (l > r && l <= N && r <= N) {
                Tree temp = node.left;
                node.left = node.right;
                node.right = temp;
                l = r;
            } else if (l == N + 1 || r == N + 1) {
                if ((l != N + 1 && node.val < l) || (r != N + 1 && node.val > r)) {
                    Tree temp = node.left;
                    node.left = node.right;
                    node.right = temp;
                    l = r;
                }
            }
            node.minVal = (l == N+1 ? node.val : l);
        }
    }

    public static void printTree(Tree t) {
        Stack<Tree> stack = new Stack<>();
        Tree curr = t;
        boolean first = true;

        while (!stack.isEmpty() || curr != null) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            if (first) {
                System.out.print(curr.val);
                first = false;
            } else {
                System.out.print(" " + curr.val);
            }
            curr = curr.right;
        }
    }

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(new File(args[0]))) {
            if (!scanner.hasNextLong()) {
                System.out.println("Error reading file");
                return;
            }
            long N = scanner.nextLong();
            Tree t = mkTree(scanner);
            arrange(t, N);
            printTree(t);
            System.out.println();
        } catch (FileNotFoundException e) {
            System.out.println("Error opening file");
        }
    }
}
